### PR TITLE
feat: handle accounts loading gracefully

### LINKED ISSUE #7419: Gracefully handle initial account loading state
### Overview

On initial login on a fresh cache it takes a moment to load all of the connected accounts, and so we end up showing confusing labels and warnings in various components.

### References and additional details

**Initial load when connected to a Native wallet**

Swapper

Screenshot 2024-07-25 at 4 29 07 PM

rFox staking

Screenshot 2024-07-25 at 4 29 31 PM

### Acceptance Criteria

We should better handle this loading state, as it can in some cases be quite lengthy (I've seen it take up to one minute).

### Need By Date

_No response_

### Screenshots/Mockups

_No response_

### Estimated effort

_No response_

### PR DESCRIPTION
**Pull request by NeOMakinG**
## Description

- I added a prop to the portfolio slice: `isAccountLoading`
- This prop is turned to true at the start of the account discovery effect, then turned false when it ends
- Consumed it in the swapper and rfox to disable the button and adapt the text if we are actually running account discovery

## Issue (if applicable)

closes #7419

## Risk
Let's say high risk because swapper button is disabled in case this is buggy

> High Risk PRs Require 2 approvals

> What protocols, transaction types, wallets or contract interactions might be affected by this PR?

## Testing
- Go to Swapper (and rFOX)
- Clear you chrome cache manually and refresh the page
- Import a native wallet
- check the button of the swapper and stake component of rFOX, it should be disabled and showing `Loading Accounts`

### Engineering
n/a

### Operations
n/a
- [ ] :checkered_flag: My feature is behind a flag and doesn't require operations testing (yet)

## Screenshots (if applicable)

rFOX
https://jam.dev/c/e28afd3a-84dd-4d7f-9630-72534531b49b

Swapper
https://jam.dev/c/ce72a79e-e372-4cea-bdba-0926d26b005c

### CODE REVIEW

File: src/components/MultiHopTrade/components/TradeInput/components/ConfirmSummary.tsx
Comment (gomesalexandre) [inline src/components/MultiHopTrade/components/TradeInput/components/ConfirmSummary.tsx:93]: blocking: plural accounts vernacular, or alternatively `isAnyAccountIdsLoading`
```suggestion
  const isAccounstLoading = useAppSelector(selectIsAccountLoading)
```

File: src/state/slices/portfolioSlice/portfolioSliceCommon.ts
Comment (gomesalexandre) [inline src/state/slices/portfolioSlice/portfolioSliceCommon.ts:61]: q: can we leverage RTK query slice instead? 

see e.g:

https://github.com/shapeshift/web/blob/d72c8d499e3898caffbc695e18efa56a0eeb701f/src/state/slices/txHistorySlice/selectors.ts#L34-L38

File: src/assets/translations/en/main.json
Comment (gomesalexandre) [inline src/assets/translations/en/main.json:158]: suggestion: possibly nicer copy for users
```suggestion
    "accountsLoading": "Accounts are currently loading",
```

Comment (NeOMakinG) [reply src/assets/translations/en/main.json:158]: Might be too big for a button, resulting in a 2 lines buttons, specially on mobile, wdyt?

File: src/components/MultiHopTrade/components/TradeInput/components/ConfirmSummary.tsx
Comment (gomesalexandre) [inline src/components/MultiHopTrade/components/TradeInput/components/ConfirmSummary.tsx:291]: preferably-blocking: with the current implementation, this will *not*  display the manual address entry until *all* accounts have loaded.

It may take some significant time before *all* accounts are loaded, and users will be stuck in "Accounts Loading" state in the meantime, despite actually having an AccountId for the requested chain in their portfolio.
  suggestion: check if there is an account for chain first, and only fallback to accounts loading checks otherwise vs. consistently using it as a guard
  
```diff
diff --git a/src/components/MultiHopTrade/components/TradeInput/components/ConfirmSummary.tsx b/src/components/MultiHopTrade/components/TradeInput/components/ConfirmSummary.tsx
index 26cdefd39a..e04f616d13 100644
--- a/src/components/MultiHopTrade/components/TradeInput/components/ConfirmSummary.tsx
+++ b/src/components/MultiHopTrade/components/TradeInput/components/ConfirmSummary.tsx
@@ -117,6 +117,14 @@ export const ConfirmSummary = ({
     buyAsset.chainId,
     buyAsset.assetId,
   ])
+
+  const displayManualAddressEntry = useMemo(() => {
+    // If accounts have finished loading, *and* there is no runtime chain support for the buy asset, display manual address entry
+    if (!walletSupportsBuyAssetChain) return !isAccountLoading
+
+    return disableThorNativeSmartContractReceive === true
+  }, [walletSupportsBuyAssetChain, disableThorNativeSmartContractReceive, isAccountLoading])
+
   const quoteHasError = useMemo(() => {
     if (!isAnyTradeQuoteLoaded) return false
     return !!activeQuoteErrors?.length || !!quoteRequestErrors?.length
@@ -285,11 +293,7 @@ export const ConfirmSummary = ({
           component={RecipientAddress}
         />
         <WithLazyMount
-          shouldUse={
-            Boolean(
-              !walletSupportsBuyAssetChain || disableThorNativeSmartContractReceive === true,
-            ) && !isAccountLoading
-          }
+          shouldUse={displayManualAddressEntry}
           shouldForceManualAddressEntry={disableThorNativeSmartContractReceive}
           component={ManualAddressEntry}
           description={manualAddressEntryDescription}
```

File: src/context/AppProvider/AppContext.tsx
Comment (gomesalexandre) [inline src/context/AppProvider/AppContext.tsx:103]: q: can `getAccount` handle this internally?

File: src/state/slices/portfolioSlice/selectors.ts
Comment (gomesalexandre) [inline src/state/slices/portfolioSlice/selectors.ts:1111]: See https://github.com/shapeshift/web/pull/7629/files#r1732559484

**Review by gomesalexandre — COMMENTED** [review]
First pass, looking sane overall, though added a few comments re: keeping things at implementation time vs. having to remember doing them at consumer level, as well as a UX improvement suggestion

File: src/assets/translations/en/main.json
Comment (NeOMakinG) [reply src/assets/translations/en/main.json]: I tested, looks ok, let's try this one and change later if needed, updated in [2d57d12](https://github.com/shapeshift/web/pull/7629/commits/2d57d12de291dbebf2c14ad97f3338aff5d7b571)

File: src/components/MultiHopTrade/components/TradeInput/components/ConfirmSummary.tsx
Comment (NeOMakinG) [reply src/components/MultiHopTrade/components/TradeInput/components/ConfirmSummary.tsx]: Cool! Why not, I did apply the diff in [2d57d12](https://github.com/shapeshift/web/pull/7629/commits/2d57d12de291dbebf2c14ad97f3338aff5d7b571)

File: src/state/slices/portfolioSlice/portfolioSliceCommon.ts
Comment (NeOMakinG) [reply src/state/slices/portfolioSlice/portfolioSliceCommon.ts]: After testing, I would say no:

- Pending queries starts 6 secs later than the actual implementation, resulting in a wrong UI first
- Pending queries are triggered absolutely every time we forceRefresh any accounts, resulting in way more loads than needed even if we already have the accounts in the cache

While it might not be the best from an engineering perspective, the actual implementation looks better to me from a UX point of view

File: src/state/slices/portfolioSlice/selectors.ts
Comment (NeOMakinG) [reply src/state/slices/portfolioSlice/selectors.ts]: See https://github.com/shapeshift/web/pull/7629/files#r1732802739

File: src/context/AppProvider/AppContext.tsx
Comment (NeOMakinG) [reply src/context/AppProvider/AppContext.tsx]: Unfortunately it looks like it's starting way too late by doing this and so the UI is displaying first that the wallet doesn't support this chain before the loading state

Comment (gomesalexandre) [reply src/context/AppProvider/AppContext.tsx]: Discussed this one out, agreed to keep the current implementation for now

File: src/state/slices/portfolioSlice/portfolioSliceCommon.ts
Comment (gomesalexandre) [reply src/state/slices/portfolioSlice/portfolioSliceCommon.ts]: Discussed this one out, agreed to keep the current implementation for now

**Review by gomesalexandre — COMMENTED** [review]
Diff conceptually looks good now! Runtime testing to follow.

**Review by gomesalexandre — APPROVED** [review]
Tested locally, confirmed this does what it says on the box:

Swapper:

https://jam.dev/c/1f5cdec5-eeed-434d-bb26-91ab7b32227f

RFOX:

https://jam.dev/c/4c8c6c1a-fa54-4c42-a15b-33c14d3a92f3

@NeOMakinG noticing a small visual inconsitency in the button colorScheme between swapper / rFOX if we want to tackle that before merging.

File: src/pages/RFOX/components/Stake/StakeInput.tsx
Comment (gomesalexandre) [inline src/pages/RFOX/components/Stake/StakeInput.tsx:553]: This is different from the blue colorScheme in swapper @NeOMakinG we may want to revert this bit

File: src/components/MultiHopTrade/components/TradeInput/components/ConfirmSummary.tsx
Comment (kaladinlight) [inline src/components/MultiHopTrade/components/TradeInput/components/ConfirmSummary.tsx:93]: can add `isAccountsLoading` to `shouldDisablePreviewButton` for posterity/paranoia

File: src/assets/translations/en/main.json
Comment (kaladinlight) [inline src/assets/translations/en/main.json:158]: `Loading Accounts` is referenced on the issue and more concise for button text imo

File: src/components/MultiHopTrade/components/TradeInput/components/ConfirmSummary.tsx
Comment (kaladinlight) [reply src/components/MultiHopTrade/components/TradeInput/components/ConfirmSummary.tsx:93]: also nitpick `isLoadingAccounts` vernacular reads better imo

File: src/components/MultiHopTrade/components/TradeInput/components/ConfirmSummary.tsx
Comment (kaladinlight) [inline src/components/MultiHopTrade/components/TradeInput/components/ConfirmSummary.tsx:125]: nitpick suggestion - more explicit and readable conditions imo. can make the wallet + sc check more concise if you want, but most important imo is the early accounts loading return.
```suggestion
  
  const displayManualAddressEntry = useMemo(() => {
    if (isAccountsLoading) return false

    if (!walletSupportsBuyAssetChain) return true
    if (disableThorNativeSmartContractReceive) return true

    return false
  }, [walletSupportsBuyAssetChain, disableThorNativeSmartContractReceive, isAccountsLoading])
```

File: src/context/AppProvider/AppContext.tsx
Comment (kaladinlight) [inline src/context/AppProvider/AppContext.tsx:125]: We could probably clean this up with a try/finally and just have the one call to set to false in the finally.

This first `hasManagedAccounts` condition also doesn't await the `getAccount.initiate` so to make a distinction between portfolio and accountMetatdata we could clarify as `isLoadingAccountMetadata` as an option

File: src/state/slices/portfolioSlice/portfolioSliceCommon.ts
Comment (kaladinlight) [inline src/state/slices/portfolioSlice/portfolioSliceCommon.ts:61]: with either `isAccountMetadataLoading` or `isPortfolioLoading` the intention would be pretty clear without any need for comment imo

File: src/components/MultiHopTrade/components/TradeInput/components/ConfirmSummary.tsx
Comment (NeOMakinG) [reply src/components/MultiHopTrade/components/TradeInput/components/ConfirmSummary.tsx]: Fixed in [e72bded](https://github.com/shapeshift/web/pull/7629/commits/e72bded769397fe997646cd23ed7ff15a20d7bc6)

File: src/components/MultiHopTrade/components/TradeInput/components/ConfirmSummary.tsx
Comment (NeOMakinG) [reply src/components/MultiHopTrade/components/TradeInput/components/ConfirmSummary.tsx]: Fixed in [e72bded](https://github.com/shapeshift/web/pull/7629/commits/e72bded769397fe997646cd23ed7ff15a20d7bc6)

File: src/context/AppProvider/AppContext.tsx
Comment (NeOMakinG) [reply src/context/AppProvider/AppContext.tsx]: You are right, used a try/finally and updated everything to `isAccountMetadataLoading` in [e72bded](https://github.com/shapeshift/web/pull/7629/commits/e72bded769397fe997646cd23ed7ff15a20d7bc6)

File: src/state/slices/portfolioSlice/portfolioSliceCommon.ts
Comment (NeOMakinG) [reply src/state/slices/portfolioSlice/portfolioSliceCommon.ts]: Removed in [e72bded](https://github.com/shapeshift/web/pull/7629/commits/e72bded769397fe997646cd23ed7ff15a20d7bc6)

File: src/pages/RFOX/components/Stake/StakeInput.tsx
Comment (kaladinlight) [reply src/pages/RFOX/components/Stake/StakeInput.tsx]: While this does technically fix the issue, it does feel weird having the account metadata loading state being tied to the validity of a wallet. Not going to block on it and not a change for this PR, but maybe something to think about regarding use/props for `ButtonWalletPredicate`.

**Review by kaladinlight — APPROVED** [review]
Looking gucci! :beers:

**Pull request merged**
